scipy/stats/stats.py

```diff
@@ -4153,6 +4154,177 @@ def binary_search(n, n1, n2, side):
+        non-negative
+
+    Returns
+    -------
+    score_statistic : float
+        Z statistic with pooled (Score) or unpooled (Wald) variance, depending
+        on the user choice of `POOLED` param.
+    p_value : float
+        P-value, the probability of obtaining a distribution at least as
+        extreme as the one that was actually observed, assuming that the
+        null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    References
+    ----------
+    .. [1] `Mehta, Cyrus & Senchaudhuri, Pralay. (2003). Conditional
+    versus Unconditional Exact Tests for Comparing Two Binomials`__.
+
+    .. __: https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf
+
+    Examples
+    --------
+    examples from [1]_.
+
+    Consider the following example of a vaccine efficacy study (Chan, 1998).
+    In a randomized clinical trial of 30 subjects, 15 were innoculated with a
+    recombinant DNA influenza vaccine and the 15 were innoculated with a
+    placebo. Twelve of the 15 subjects in the placebo group (80%) eventually
+    became infected with influenza whereas for the vaccine group, only 7 of
+    the 15 subjects (47%) became infected. The data are tabulated as a 2 × 2
+    table::
+
+                    Vaccine  Placebo
+        Yes             8        2
+        No              1        5
+
+    We use this table to find the p-value:
+    >>> import scipy.stats as stats
+    >>> TX, pvalue = stats.barnard_exact([[7, 12], [8, 3]])
+    >>> pvalue
+    0.0681...
+
+    """
+    nuisance_num = 100
+
+    if num_it < 0:
+        raise ValueError(
+            "Number of iterations must be non-negative, "
+            f"found {num_it!r}"
+        )
+
+    c = np.asarray(
+        table,
+        dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0) or 0 in c.sum(axis=1):
+        # If both values in a row or column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return np.nan, 1.0
+
+    total_c1, total_c2 = c.sum(axis=0)
+    n = total_c1 + total_c2
+
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1)
+
+    p1, p2 = x1 / total_c1, x2 / total_c2
+
+    if POOLED:
+        p = (x1 + x2) / (total_c1 + total_c2)
+        var_p1_p2 = p * (1 - p) * (1 / total_c1 + 1 / total_c2)
+    else:
+        var_p1_p2 = p1 * (1 - p1) / total_c1 + p2 * (1 - p2) / total_c2
+
+    # To avoid warning when dividing by 0
+    with np.errstate(divide='ignore', invalid='ignore'):
+        TX = (p2 - p1) / np.sqrt(var_p1_p2)
```

Comment:
Where var_p1_p2 is null, TX is NaN. Should be +-inf.

Comparing a NaN to a value is always False.

Except in rare cases, this does not change the result, but is some special cases it can occurs.